src/app/InteractionModelEngine.cpp

```diff
@@ -253,24 +258,50 @@ CHIP_ERROR InteractionModelEngine::ShutdownSubscription(const ScopedNodeId & aPe
 
 void InteractionModelEngine::ShutdownSubscriptions(FabricIndex aFabricIndex, NodeId aPeerNodeId)
 {
-    for (auto * readClient = mpActiveReadClientList; readClient != nullptr; readClient = readClient->GetNextClient())
+    // This is assuming that ReadClient::Close will not affect any other
+    // ReadClients in the list.
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr;)
     {
+        // Grab the next client now, because we might be about to delete readClient.
+        auto * nextClient = readClient->GetNextClient();
         if (readClient->IsSubscriptionType() && readClient->GetFabricIndex() == aFabricIndex &&
             readClient->GetPeerNodeId() == aPeerNodeId)
         {
             readClient->Close(CHIP_NO_ERROR);
         }
+        readClient = nextClient;
+    }
+}
+
+void InteractionModelEngine::ShutdownSubscriptions(FabricIndex aFabricIndex)
+{
+    // This is assuming that ReadClient::Close will not affect any other
+    // ReadClients in the list.
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr;)
+    {
+        // Grab the next client now, because we might be about to delete readClient.
+        auto * nextClient = readClient->GetNextClient();
+        if (readClient->IsSubscriptionType() && readClient->GetFabricIndex() == aFabricIndex)
+        {
+            readClient->Close(CHIP_NO_ERROR);
+        }
+        readClient = nextClient;
     }
 }
 
 void InteractionModelEngine::ShutdownAllSubscriptions()
 {
-    for (auto * readClient = mpActiveReadClientList; readClient != nullptr; readClient = readClient->GetNextClient())
+    // This is assuming that ReadClient::Close will not affect any other
+    // ReadClients in the list.
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr;)
     {
+        // Grab the next client now, because we might be about to delete readClient.
+        auto * nextClient = readClient->GetNextClient();
         if (readClient->IsSubscriptionType())
         {
             readClient->Close(CHIP_NO_ERROR);
         }
+        readClient = nextClient;
```

Comment:
What about having a single method with `chip::Optional` parameters for the `FabricIndex` and the `NodeId` ?
That would avoid having to duplicate the loops/comments. I'm thinking of something like:

```
void InteractionModelEngine::ShutdownSubscriptions(Optional<FabricIndex> fabricIndex = Optional<FabricIndex>::Missing(), Optional<NodeId> nodeId = Optional<NodeId>::Missing())
{
    // This is assuming that ReadClient::Close will not affect any other
     // ReadClients in the list.
     for (auto * readClient = mpActiveReadClientList; readClient != nullptr;)
     {
         // Grab the next client now, because we might be about to delete readClient.
         auto * nextClient = readClient->GetNextClient();
         if (readClient->IsSubscriptionType())
         {
             bool validFabric = !fabricIndex.HasValue() || (fabricIndex.Value) == readClient->GetFabricIndex());
             bool validNodeId = !nodeId.HasValue() || (nodeId.Value() == readClient->GetNodeId());
             if (validFabric && validNodeId)
             {
                 readClient->Close(CHIP_NO_ERROR);
             }
         }
         readClient = nextClient;
     }
}

void InteractionModelEngine::ShutdownAllSubscriptions()
{
     ShutdowSubscriptions();
}

void InteractionModelEngine::ShutdownSubscriptions(FabricIndex aFabricIndex)
{
    ShutdowSubscriptions(MakeOptional(aFabricIndex));
}

void InteractionModelEngine::ShutdownSubscriptions(FabricIndex aFabricIndex, NodeId aPeerNodeId)
{
    ShutdowSubscriptions(MakeOptional(aFabricIndex), MakeOptional(nodeId));
}
```

Reply:
Good idea, done.